config/puma/production.rb

```diff
@@ -0,0 +1,14 @@
+# Configure “min” to be the minimum number of threads to use to answer
+# requests and “max” the maximum.
+#
+# The default is “0, 16”.
+#
+threads ENV.fetch('FOREMAN_PUMA_THREADS_MIN', 0).to_i, ENV.fetch('FOREMAN_PUMA_THREADS_MAX', 16).to_i
+
+# === Cluster mode ===
+
+# How many worker processes to run.
+#
+# The default is “0” for puma. Recommending "4" for foreman
+#
+workers ENV.fetch('FOREMAN_PUMA_WORKERS', 4).to_i
```

Comment:
Given each worker spawns its own process and therefore has a memory footprint, do we need 4 as a base? How does this compare to the starting configuration of Passenger?

Reply:
I am still testing the memory footprint.

Reply:
I did some interesting testing with default passenger tuning in foreman vs. puma in foreman:

**For passenger:** 
Using default passenger tuning defined by `/usr/share/foreman-installer/config/foreman.hiera/tuning/common.yaml`
```sh
# cat /usr/share/foreman-installer/config/foreman.hiera/tuning/common.yaml | grep passenger
apache::mod::passenger::passenger_max_pool_size: 12
apache::mod::passenger::passenger_max_instances_per_app: 6
apache::mod::passenger::passenger_max_request_queue_size: 250
apache::mod::passenger::passenger_stat_throttle_rate: 120
apache::mod::passenger::passenger_max_requests: 10000
apache::mod::passenger::passenger_min_instances: 1
apache::mod::passenger::passenger_max_preloader_idle_time: 0

$ ps aux | grep Passenger
root      2977  0.0  0.0 217324  2020 ?        Ssl  22:08   0:00 PassengerWatchdog
root      2982  0.4  0.0 905228  4416 ?        Sl   22:08   0:02 PassengerHelperAgent
nobody    2996  0.0  0.0 220776  4332 ?        Sl   22:08   0:00 PassengerLoggingAgent
foreman   3129  2.2  2.5 777972 297728 ?       Sl   22:08   0:15 Passenger AppPreloader: /usr/share/foreman
foreman   3326  0.5  4.2 1562036 497708 ?      Sl   22:09   0:03 Passenger RackApp: /usr/share/foreman
```
*Total memory: 3683336K or 3.6G*

**For Puma:**
Note: Default threads min and max for puma is 0,16. We will continue to use the same.

Using `4 workers`:
```sh
$ ps aux | grep puma
foreman   7052  0.0  2.7 1042244 322044 ?      Sl   Feb11   0:33 puma 3.11.4 (tcp://127.0.0.1:3000) [foreman]
foreman   7094  0.0  4.4 1311536 527564 ?      Sl   Feb11   0:17 puma: cluster worker 0: 7052 [foreman]
foreman   7103  0.0  4.2 1311536 499152 ?      Sl   Feb11   0:18 puma: cluster worker 1: 7052 [foreman]
foreman   7112  0.0  4.8 1377072 570580 ?      Sl   Feb11   0:16 puma: cluster worker 2: 7052 [foreman]
foreman   7121  0.1  4.1 1606920 487732 ?      Sl   Feb11   1:43 puma: cluster worker 3: 7052 [foreman]
```
*Toal memory usage: 6649308 or 6.6G*

Using `2 workers`:
```sh
$ ps aux | grep puma
root      2548  0.0  0.0  26168  1972 pts/0    S+   16:45   0:00 vi /usr/share/foreman/config/puma/production.rb
foreman  17427 26.8  2.5 975940 297524 ?       Sl   22:40   0:15 puma 3.11.4 (tcp://127.0.0.1:3000) [foreman]
foreman  17493  0.8  2.5 1114124 307896 ?      Sl   22:40   0:00 puma: cluster worker 0: 17427 [foreman]
foreman  17502  7.9  3.8 1564344 454620 ?      Sl   22:40   0:03 puma: cluster worker 1: 17427 [foreman]
```
*Total memory usage: 3654408 or 3.6G*



**Result**
Using `2 workers` in puma results in the similar memory consumption of default passenger settings in Foreman. So I am recommending **2 workers** for puma, as default configuration in foreman.

Reply:
@ehelms @ekohl ping to know your thoughts about my recommendation.

Reply:
Given that info, I like 2 workers. It also helps enforce testing of a clustered approach. Devel looks like it does 5 (which I don't know why its that high):

https://github.com/theforeman/puppet-katello_devel/blob/master/manifests/params.pp#L52

We might should align these two or add a development config file similar to the production one.

Reply:
Created https://github.com/theforeman/puppet-katello_devel/pull/226 to update puma workers to 2 by default.